From 40cde9dcc581bfde0e888dbe048a3774689f9bff Mon Sep 17 00:00:00 2001
From: Swee Yee Fonn <swee.yee.fonn@intel.com>
Date: Wed, 4 Nov 2020 20:24:17 +0800
Subject: [PATCH] Support keybox provisioning over vsock

Update script to delete remnant RPMB_DATA when flashing
new android image.
Update android launch script to start keybox
provisioning daemon when detect first boot.
Update setup-host script to install tpm2-tools
and tpm2-tss required for decrypting keybox.

Tracked-on: OAM-9558
Change-Id: I8034a646bc31161053339b4b7941737d4ac71fef
Signed-off-by: Swee Yee Fonn <swee.yee.fonn@intel.com>
---
 scripts/setup_host.sh      | 14 ++++++++++++++
 scripts/start_civ.sh       |  6 ++++++
 scripts/start_flash_usb.sh |  4 ++++
 3 files changed, 24 insertions(+)

diff --git a/scripts/setup_host.sh b/scripts/setup_host.sh
index f07b89c..b8fad50 100755
--- a/scripts/setup_host.sh
+++ b/scripts/setup_host.sh
@@ -110,6 +110,19 @@ function ubu_build_ovmf_gvt(){
     cd -
 }
 
+function ubu_install_tss2(){
+    sudo apt install -y tpm2-tools
+    sudo apt install -y git autoconf-archive libcmocka0 libcmocka-dev procps iproute2 build-essential git pkg-config gcc libtool automake libssl-dev uthash-dev autoconf doxygen libjson-c-dev libini-config-dev libgcrypt-dev libcurl4-openssl-dev
+    git clone https://github.com/tpm2-software/tpm2-tss
+    cd tpm2-tss
+    git checkout 2.4.5 -b build
+    ./bootstrap
+    ./configure
+    make -j4
+    sudo make install
+    cd -
+}
+
 function ubu_enable_host_gvt(){
     if [[ ! `cat /etc/default/grub` =~ "i915.enable_gvt=1 intel_iommu=on i915.force_probe=*" ]]; then
         read -p "The grub entry in '/etc/default/grub' will be updated for enabling GVT-g and GVT-d, do you want to continue? [Y/n]" res
@@ -400,6 +413,7 @@ ubu_enable_additional_kernel_params
 prepare_required_scripts
 setup_sof
 ubu_install_swtpm
+ubu_install_tss2
 ubu_update_bt_fw
 
 ask_reboot
diff --git a/scripts/start_civ.sh b/scripts/start_civ.sh
index 6edcd86..0149e57 100644
--- a/scripts/start_civ.sh
+++ b/scripts/start_civ.sh
@@ -22,6 +22,7 @@ GUEST_KIRQ_CHIP="-machine kernel_irqchip=on"
 GUEST_VGA_DEV="-device virtio-gpu-pci"
 GUEST_RPMB_DEV=
 GUEST_RPMB_DEV_PID=
+GUEST_KEYBOX_PROVISIOND_PID=
 GUEST_RPMB_DEV_SOCK=
 GUEST_THERMAL_DAEMON_PID=
 GUEST_BATTERY_DAEMON_PID=
@@ -126,11 +127,15 @@ function kill_daemon_proc() {
 
 function setup_rpmb_dev() {
     local RPMB_DEV=$SCRIPTS_DIR/rpmb_dev
+    local KEYBOX_PROVISIOND=$SCRIPTS_DIR/keybox_provisiond
     local RPMB_DATA=$WORK_DIR/RPMB_DATA
     GUEST_RPMB_DEV_SOCK=$WORK_DIR/rpmb_sock
     # RPMB_DATA is created and initialized with specific key, if this file
     # is deleted by accidently, create a new one without any data.
     if [ ! -f $RPMB_DATA ]; then
+        echo "starting keybox_provisiond"
+        setsid $KEYBOX_PROVISIOND -p 5392 -t 0x81000004 > $WORK_DIR/keybox_provisiond.log  2>&1 &
+        GUEST_KEYBOX_PROVISIOND_PID=$!
         echo "Creating RPMB DATA..."
         $RPMB_DEV --dev $RPMB_DATA --init --size 2048
     fi
@@ -164,6 +169,7 @@ function setup_rpmb_dev() {
 function cleanup_rpmb_dev() {
     kill_daemon_proc "$GUEST_RPMB_DEV_PID" "rpmb_dev"
     [[ -z "$GUEST_RPMB_DEV_SOCK" ]] || ( [[ -S $GUEST_RPMB_DEV_SOCK ]] && rm $GUEST_RPMB_DEV_SOCK )
+    kill_daemon_proc "$GUEST_KEYBOX_PROVISIOND_PID" "keybox_provisiond"
 }
 
 function setup_swtpm() {
diff --git a/scripts/start_flash_usb.sh b/scripts/start_flash_usb.sh
index d1801f1..b2da6ac 100755
--- a/scripts/start_flash_usb.sh
+++ b/scripts/start_flash_usb.sh
@@ -88,6 +88,10 @@ else
     read option
     if [ "$option" == 'y' ]; then
       rm $GUEST_IMAGE
+      if [ -f RPMB_DATA ]
+      then
+        rm RPMB_DATA
+      fi
     else
       exit 1
     fi
-- 
2.17.1

